### DISCUSSION TITLE
Can't set a Custom UDP Transport to set a interface while list in ROS2 Humble container

### DISCUSSION THREAD

**Discussion by ghost** [eProsima/Fast-DDS#3737] [Support]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

I'm trying to run the ROS2 minimal publisher examples inside a `ros:humble-ros-base` container, by setting a custom UDP Transport to only work with a specific network interface.

I'm setting these env variables:

``` bash
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
export FASTRTPS_DEFAULT_PROFILES_FILE=/fastdds_profiles/profile1.xml
export RMW_FASTRTPS_USE_QOS_FROM_XML=1
```

And then run one of the following commands:

Python:

``` bash
ros2 run examples_rclpy_minimal_publisher publisher_member_function
```

Cpp:

``` bash
ros2 run examples_rclcpp_minimal_publisher publisher_member_function
```

I'd expect the nodes just to normally publish.

### Current behavior

None of them work.

The Python example prints this:

```

>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'rcl node's rmw handle is invalid, at ./src/rcl/node.c:415'

with this new error message:

  'Unable to fini publisher for node., at ./src/rcl/node.c:380'

rcutils_reset_error() should be called after error handling to avoid this.
<<<

>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'Unable to fini publisher for node., at ./src/rcl/node.c:380'

with this new error message:

  'node argument is null, at ./src/rmw_node.cpp:80'

rcutils_reset_error() should be called after error handling to avoid this.
<<<

>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'node argument is null, at ./src/rmw_node.cpp:80, at ./src/rcl/node.c:386'

with this new error message:

  'guard_condition argument is null, at ./src/rcl/guard_condition.c:126'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/node.py:175: RuntimeWarning: Failed to fini node: guard_condition argument is null, at ./src/rcl/guard_condition.c:126, at ./src/rcl/node.c:391
  self.__node = _rclpy.Node(
Traceback (most recent call last):
  File "/opt/ros/humble/lib/examples_rclpy_minimal_publisher/publisher_member_function", line 33, in <module>
    sys.exit(load_entry_point('examples-rclpy-minimal-publisher==0.15.1', 'console_scripts', 'publisher_member_function')())
  File "/opt/ros/humble/lib/python3.10/site-packages/examples_rclpy_minimal_publisher/publisher_member_function.py", line 41, in main
    minimal_publisher = MinimalPublisher()
  File "/opt/ros/humble/lib/python3.10/site-packages/examples_rclpy_minimal_publisher/publisher_member_function.py", line 24, in __init__
    super().__init__('minimal_publisher')
  File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/node.py", line 175, in __init__
    self.__node = _rclpy.Node(
RuntimeError: Not enough memory in the buffer stream
[rcl|context.c:157] failed to finalize rmw context while cleaning up context, memory may be leaked: Finalizing a context with active nodes, at ./src/rmw_init.cpp:159
sys:1: RuntimeWarning: [rclpy| "./src/rclpy/context.cpp" : 83 ]: failed to fini rcl_context_t: error not set
[ros2run]: Segmentation fault
```

And the CPP prints this:

```
terminate called after throwing an instance of 'eprosima::fastcdr::exception::NotEnoughMemoryException'
  what():  Not enough memory in the buffer stream
[ros2run]: Aborted
```

If I set `RMW_FASTRTPS_USE_QOS_FROM_XML=0` I stopped getting the errors, but it publish through all the network interfaces (so I guess it doesn't consider the XML).

## Files

Create an empty working directory, and there put these files:

### Dockerfile

Path: `dockerfiles/base.dockerfile`

``` Dockerfile
FROM ros:humble-ros-base

RUN apt-get update && \
    apt-get install -y \
        net-tools \
    && rm -rf /var/lib/apt/lists/* \
    && apt-get clean

RUN apt-get update && \
    apt-get install -y \
        ros-humble-examples-rclpy-minimal-publisher \
        ros-humble-examples-rclpy-minimal-subscriber \
        ros-humble-examples-rclpy-minimal-client \
        ros-humble-examples-rclpy-minimal-service \
        ros-humble-examples-rclcpp-minimal-publisher \
        ros-humble-examples-rclcpp-minimal-subscriber \
        ros-humble-examples-rclcpp-minimal-client \
        ros-humble-examples-rclcpp-minimal-service \
    && rm -rf /var/lib/apt/lists/* \
    && apt-get clean
```

### Docker compose

Path: `docker-compose.yml`

``` yaml
version: '3.3'

services:
  base:
    image: mycompany/test/distributed/base
    build:
      context: .
      dockerfile: dockerfiles/base.dockerfile
    volumes:
      - ./mount/fastdds_profiles:/fastdds_profiles
    command: tail -F anything
    networks:
      robot:
        ipv4_address: 10.5.0.2
      apps:
        ipv4_address: 10.6.0.2
    

networks:
  robot:
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 10.5.0.0/16
          gateway: 10.5.0.1
    driver_opts:
      com.docker.network.bridge.name: robot
  apps:
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 10.6.0.0/16
          gateway: 10.6.0.1
    driver_opts:
      com.docker.network.bridge.name: apps
```

### Profile XML file

Path: `mount/fastdds_profiles/profile1.xml`

Content: In the "XML configuration file" section.

## Run

Up the container:

``` bash
docker compose up
```

It'll run the container in background, now open a new terminal:

``` bash
docker exec -it issue-base-1 bash
```

Set environment:

``` bash
source /opt/ros/humble/setup.bash

export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
export FASTRTPS_DEFAULT_PROFILES_FILE=/fastdds_profiles/profile1.xml
export RMW_FASTRTPS_USE_QOS_FROM_XML=1
```

And now try to run any of the two examples.

### Fast DDS version/commit

The one that comes with the docker image. When running:

``` bash
dpkg -l | grep fastrt
```

I get:

```
ii  ros-humble-fastrtps                               2.6.5-1jammy.20230623.031537            amd64        *eprosima Fast DDS* (formerly Fast RTPS) is a C++ implementation of the DDS (Data Distribution Service) standard of the OMG (Object Management Group).
ii  ros-humble-fastrtps-cmake-module                  2.2.0-2jammy.20230623.030445            amd64        Provide CMake module to find eProsima FastRTPS.
ii  ros-humble-rmw-fastrtps-cpp                       6.2.2-1jammy.20230623.042827            amd64        Implement the ROS middleware interface using eProsima FastRTPS static code generation in C++.
ii  ros-humble-rmw-fastrtps-shared-cpp                6.2.2-1jammy.20230623.041808            amd64        Code shared on static and dynamic type support of rmw_fastrtps_cpp.
ii  ros-humble-rosidl-typesupport-fastrtps-c          2.2.0-2jammy.20230623.035604            amd64        Generate the C interfaces for eProsima FastRTPS.
ii  ros-humble-rosidl-typesupport-fastrtps-cpp        2.2.0-2jammy.20230623.035243            amd64        Generate the C++ interfaces for eProsima FastRTPS.
```

### Platform/Architecture

Other. Please specify in Additional context section.

### Transport layer

UDPv4

### Additional context

* OS: Ubuntu 22.04.2 LTS
* Container base image: ros:humble-ros-base
* Docker version 24.0.2, build cb74dfc
* Docker Compose version v2.18.1

### XML configuration file

```xml
<?xml version="1.0" encoding="UTF-8" ?>

<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">

    <transport_descriptors>
        <transport_descriptor>
            <transport_id>CustomUDPTransport</transport_id>
            <type>UDPv4</type>
            <interfaceWhiteList>
                <address>lo</address>
            </interfaceWhiteList>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="participant_profile_ros2">
        <rtps>
            <useBuiltinTransports>false</useBuiltinTransports>
            <userTransports>
                <transport_id>CustomUDPTransport</transport_id>
            </userTransports>
        </rtps>
    </participant>

</profiles>
```

### Relevant log output

```shell
Included in "Current Behavior"
```

**Comment by JLBuenoLopez** [comment]
Thanks for your ticket @camilocamachour 

There are 2 issues in your ticket:

1. `RMW_FASTRTPS_USE_QOS_FROM_XML` environment variable has to be set only when the Memory Management Policy and/or Publication mode is to be configured through the XML configuration file instead of using the RMW default configuration (as explained in the [RMW FastRTPS README](https://github.com/ros2/rmw_fastrtps#full-qos-configuration)). Consequently, in your use case, you do not need to modify them. The Fast CDR exception (`Not enough memory in the buffer stream`) that is thrown is due to this. Fast DDS 2.6.x (which is the Fast DDS version running in ROS 2 Humble) has as default [Memory Management Policy](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#memorymanagementpolicy) `PREALLOC` while the RMW sets the default value to `PREALLOC_WITH_REALLOC`. As ROS 2 most common types are unbounded, when the preallocated memory is filled up, the exception you are seeing will be thrown.
2. In order to whitelist a specific interface you cannot use the DNS name. Even though this is supported when defining locators, Fast DDS does not recognize names when setting the whitelisted interfaces. Try instead of setting `lo`, setting `127.0.0.1`. Also, localhost is treated differently in Fast DDS, because that interface is only initialized if no other interface is found. Consequently, you will need to setup the whitelist both in the publisher and subscriber for the specific case of the localhost interface.

Please, give it a try and consider closing the issue if it is solved.

**Comment by ghost** [comment]
I changed my xml file to:

```

<?xml version="1.0" encoding="UTF-8" ?>

<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">

    <transport_descriptors>
        <transport_descriptor>
            <transport_id>CustomUDPTransport</transport_id>
            <type>UDPv4</type>
            <interfaceWhiteList>
                <address>127.0.0.1</address>
            </interfaceWhiteList>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="participant_profile_ros2">
        <rtps>
            <useBuiltinTransports>false</useBuiltinTransports>
            <userTransports>
                <transport_id>CustomUDPTransport</transport_id>
            </userTransports>
        </rtps>
    </participant>

</profiles>
```

and didn't set the `RMW_FASTRTPS_USE_QOS_FROM_XML` variable.

The examples run but I can see/echo the published topic from other containers that are connected to the networks `robot` and `apps`. So it's not actually limiting the publishing to only the local host network.

**Comment by ghost** [comment]
Also, I have two bridged networks in the container: `robot` (IP `10.5.0.2`) and `apps` (IP `10.6.0.2`). If I set:

``` xml
            <interfaceWhiteList>
                <address>10.5.0.2</address>
            </interfaceWhiteList>
```

I still see the topic in containers connected to any of the networks, not only in the ones connected to `robot` (as I guess it should happen).

**Comment by JLBuenoLopez** [comment]
I do not think the profile is being loaded into the DomainParticipant. Fast DDS provides API to create a DomainParticipant using directly a XML profile but this is not exposed into ROS 2. The only way to load the participant profile is setting the profile as the default one using the attribute `is_default_profile` set to `"true"`. DataReaders/DataWriters (or Publishers/Subscribers/Services/Clients in ROS 2) can have its own profile if the profile name is called as the Topic in which they are created. Let me know if this works for you.

**Comment by JLBuenoLopez** [comment]
@camilocamachour, I am going to close the issue. Please, feel free to reopen if you are still having problems.

**Comment by ghost** [comment]
This is the network schema I'm using:

![image](https://github.com/eProsima/Fast-DDS/assets/89929953/115ac26f-b587-40dc-9b8a-820cb98b5ff5)

I added the `is_default_profile="true"` attribute to the participant and this is the behavior:

### Setting whitelist for localhost

This is the profile file:

``` xml

<?xml version="1.0" encoding="UTF-8" ?>

<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">

    <transport_descriptors>
        <transport_descriptor>
            <transport_id>CustomUDPTransport</transport_id>
            <type>UDPv4</type>
            <interfaceWhiteList>
                <address>127.0.0.1</address>
            </interfaceWhiteList>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="participant_profile_ros2" is_default_profile="true">
        <rtps>
            <useBuiltinTransports>false</useBuiltinTransports>
            <userTransports>
                <transport_id>CustomUDPTransport</transport_id>
            </userTransports>
        </rtps>
    </participant>

</profiles>
```

I publish from the container `base`, and I don't see the topic in either the same container or other containers. I made sure the profile env variables are set in both the publisher and the subscriber terminals.

### Setting the whitelist for `apps` network

``` xml

<?xml version="1.0" encoding="UTF-8" ?>

<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">

    <transport_descriptors>
        <transport_descriptor>
            <transport_id>CustomUDPTransport</transport_id>
            <type>UDPv4</type>
            <interfaceWhiteList>
                <address>10.6.0.2</address>
                <address>10.6.0.4</address>
            </interfaceWhiteList>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="participant_profile_ros2" is_default_profile="true">
        <rtps>
            <useBuiltinTransports>false</useBuiltinTransports>
            <userTransports>
                <transport_id>CustomUDPTransport</transport_id>
            </userTransports>
        </rtps>
    </participant>

</profiles>
```

I publish from base and I do see the topic in the same container, but I don't see it from the `app` container. I made sure I have the same profile in both containers.

**Comment by JLBuenoLopez** [comment]
I am moving the discussion to the support discussion forum as per [Fast DDS CONTRIBUTING guidelines](https://github.com/eProsima/Fast-DDS/blob/master/CONTRIBUTING.md). We will try to reproduce the issue and get back to you.

**Comment by juanlofer-eprosima** [comment]
Hi @camilocamachour , I just wanted to mention that the two statements written in the second point of https://github.com/eProsima/Fast-DDS/discussions/3737#discussioncomment-6505829 no longer hold for recent versions of Fast-DDS. Also, feel free to check the recently introduced [netmask filtering](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/interfaces.html#netmask-filtering) feature, in case it may be of any use for your scenario.

Cheers!
